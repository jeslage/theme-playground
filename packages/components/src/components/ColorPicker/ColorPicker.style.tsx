import styled from 'styled-components';

const StlyedColorPicker = styled.div`
  display: flex;
  align-items: center;
  width: 100%;

  .colorPicker__cover {
    position: fixed;
    top: 0px;
    right: 0px;
    bottom: 0px;
    left: 0px;
    outline: none;
    border: none;
    background: none;
  }

  .colorPicker__open {
    outline: none;
    border: none;
    cursor: pointer;
    background: none;
    padding: 5px;
    width: 60px;
    border-radius: 30px;
    border: 1px solid ${({ theme }) => theme.border};

    span {
      display: block;
      border-radius: 30px;
      height: 20px;
    }
  }

  .colorPicker__wrapper {
    position: relative;
  }

  .colorPicker__content {
    position: absolute;
    right: 0;
    top: 40px;
    z-index: 1;
    border-radius: 5px;
    overflow: hidden;
    box-shadow: rgba(0, 0, 0, 0.15) 0px 0px 2px, rgba(0, 0, 0, 0.15) 0px 4px 8px;
  }
`;

export default StlyedColorPicker;
