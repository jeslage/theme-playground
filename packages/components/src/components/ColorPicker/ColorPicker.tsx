import * as React from 'react';
import { ChromePicker } from 'react-color';

import StyledColorPicker from './ColorPicker.style';
import Label from '../Label/Label';

export interface Props {
  iconBefore?: JSX.Element;
  description?: string;
  label: string;
  onChange: (hex: string) => void;
  value: string;
}

const ColorPicker: React.FC<Props> = ({
  iconBefore,
  description,
  label,
  onChange,
  value
}) => {
  const [visible, setVisible] = React.useState(false);
  const content = React.useRef<HTMLDivElement>(null);

  const handleOutsideClick = e => {
    if (content.current && !content.current.contains(e.target) && visible)
      setVisible(false);
  };

  React.useEffect(() => {
    window.addEventListener('mousedown', handleOutsideClick);

    return () => {
      window.removeEventListener('mousedown', handleOutsideClick);
    };
  }, [handleOutsideClick]);

  const handleChange = ({ hex, rgb }) => {
    if (onChange) {
      let newColor = hex;

      if (rgb.a < 1) {
        newColor = `rgba(${rgb.r},${rgb.g},${rgb.b},${rgb.a})`;
      }

      onChange(newColor);
    }
  };
  return (
    <StyledColorPicker>
      <Label iconBefore={iconBefore} label={label} description={description} />

      <div className="colorPicker__wrapper">
        <button
          type="button"
          onClick={() => setVisible(prev => !prev)}
          aria-label="Open color picker"
          className="colorPicker__open"
        >
          <span style={{ background: value }} />
        </button>

        {visible && (
          <>
            <button
              type="button"
              onClick={() => setVisible(false)}
              aria-label="Close color picker"
              className="colorPicker__cover"
            />
            <div className="colorPicker__content" ref={content}>
              <ChromePicker color={value} onChangeComplete={handleChange} />
            </div>
          </>
        )}
      </div>
    </StyledColorPicker>
  );
};

export default ColorPicker;
