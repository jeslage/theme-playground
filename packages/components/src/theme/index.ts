export const darkTheme = {
  base: 'dark',
  backgroundDark: '#2f2f2f',
  backgroundLight: '#333',
  border: '#999999',
  text: '#FFFFFF'
};

export const lightTheme = {
  base: 'light',
  backgroundDark: '#f6f9fc',
  backgroundLight: '#fff',
  border: '#666666',
  text: '#333333'
};

export const getUiTheme = theme => {
  if (theme === undefined) return darkTheme;

  if (typeof theme === 'string') {
    if (theme === 'dark') {
      return darkTheme;
    } else if (theme === 'light') {
      return lightTheme;
    }

    throw Error(
      `Theme "${theme}" is not available. Please use "light" or "dark", or provide a custom theme object.`
    );
  }

  if (theme.base && theme.base === 'light') {
    return {
      ...lightTheme,
      ...theme
    };
  }

  return {
    ...darkTheme,
    ...theme
  };
};
