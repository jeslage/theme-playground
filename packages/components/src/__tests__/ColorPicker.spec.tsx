import 'jest';
import * as React from 'react';
import { shallow } from 'enzyme';

import ColorPicker from '../components/ColorPicker/ColorPicker';
import StyledColorPicker from '../components/ColorPicker/ColorPicker.style';
import { darkTheme } from '../theme';

describe('ColorPicker', () => {
  it('ColorPicker should render correctly', () => {
    const component = shallow(
      <ColorPicker
        onChange={value => console.log(value)}
        label="Label"
        description="Testbeschreibung"
        value="#000"
        iconBefore={
          <svg height="24" viewBox="0 0 24 24" width="24">
            <rect height="2" rx="1" width="12" x="6" y="11" />
          </svg>
        }
      />
    );

    expect(component).toMatchSnapshot();
  });

  it('StyledColorPicker should render correctly', () => {
    const component = shallow(
      <StyledColorPicker theme={{ colors: darkTheme }} />
    );

    expect(component).toMatchSnapshot();
  });
});
