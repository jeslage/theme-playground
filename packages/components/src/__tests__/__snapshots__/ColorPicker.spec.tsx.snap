// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorPicker ColorPicker should render correctly 1`] = `
<styled.div>
  <Label
    description="Testbeschreibung"
    iconBefore={
      <svg
        height="24"
        viewBox="0 0 24 24"
        width="24"
      >
        <rect
          height="2"
          rx="1"
          width="12"
          x="6"
          y="11"
        />
      </svg>
    }
    label="Label"
  />
  <div
    className="colorPicker__wrapper"
  >
    <button
      aria-label="Open color picker"
      className="colorPicker__open"
      onClick={[Function]}
      type="button"
    >
      <span
        style={
          Object {
            "background": "#000",
          }
        }
      />
    </button>
  </div>
</styled.div>
`;

exports[`ColorPicker StyledColorPicker should render correctly 1`] = `
<div
  className="sc-AxjAm jOQktQ"
/>
`;
